com.ca.lsp.cobol/lsp-service-cobol/src/test/java/com/ca/lsp/cobol/usecases/engine/TestPreprocessorListenerImpl.java

```diff
@@ -0,0 +1,350 @@
+  private int[] lineShifts;
+  private CommonTokenStream tokens;
+  private String documentUri;
+  private String copybookName;
+  private Map<String, Diagnostic> expectedDiagnostics;
+
+  TestPreprocessorListenerImpl(
+      CommonTokenStream tokens,
+      String documentName,
+      String documentUri,
+      int numberOfLines,
+      Map<String, Diagnostic> expectedDiagnostics) {
+    this.tokens = tokens;
+    this.documentUri = documentUri;
+    this.copybookName = documentName;
+    lineShifts = new int[numberOfLines];
+    this.expectedDiagnostics = expectedDiagnostics;
+    contexts.push(new StringBuilder());
+    diagnostics.put(documentUri, new ArrayList<>());
+    ofNullable(documentName).ifPresent(defineCopybook(documentUri));
+  }
+
+  /**
+   * Get result of processing, including cleaned-up text, semantic context and diagnostics
+   *
+   * @return a {@link TestData} of the processing
+   */
+  @Nonnull
+  TestData getProcessingResult() {
+    return new TestData(
+        peek().toString(),
+        copybookName,
+        diagnostics,
+        variableDefinitions,
+        variableUsages,
+        paragraphDefinitions,
+        paragraphUsages,
+        copybookDefinitions,
+        copybookUsages);
+  }
+
+  @Override
+  public void enterErrorStatement(ErrorStatementContext ctx) {
+    push();
+  }
+
+  @Override
+  public void exitErrorStatement(ErrorStatementContext ctx) {
+    pop();
+    ReplacementContext replacementContext =
+        ofNullable(ctx.word()).map(WordContext::replacement).orElse(null);
+    String text =
+        ofNullable(ctx.word())
+            .map(WordContext::identifier)
+            .map(RuleContext::getText)
+            .orElse(ofNullable(ctx.STRINGLITERAL()).map(ParseTree::getText).orElse(""));
+
+    processToken(text, ctx, replacementContext, null, ctx.diagnostic());
+  }
+
+  @Override
+  public void enterCopybookStatement(CopybookStatementContext ctx) {
+    push();
+  }
+
+  @Override
+  public void exitCopybookStatement(CopybookStatementContext ctx) {
+    pop();
+    ofNullable(ctx.copybookUsage())
+        .map(CopybookUsageContext::cpyIdentifier)
+        .ifPresent(processCopybook(ctx, copybookUsages));
+    ofNullable(ctx.copybookDefinition())
+        .map(CopybookDefinitionContext::cpyIdentifier)
+        .ifPresent(processCopybook(ctx, copybookDefinitions));
+  }
+
+  private Consumer<CpyIdentifierContext> processCopybook(
+      CopybookStatementContext ctx, Map<String, List<Location>> copybookUsages) {
+    return it ->
+        processToken(
+            it.cpyName().getText(), ctx, it.replacement(), copybookUsages, ctx.diagnostic());
+  }
+
+  @Override
+  public void enterVariableStatement(VariableStatementContext ctx) {
+    push();
+  }
+
+  @Override
+  public void exitVariableStatement(VariableStatementContext ctx) {
+    pop();
+    ofNullable(ctx.variableUsage())
+        .map(VariableUsageContext::word)
+        .ifPresent(
+            it ->
+                processToken(
```

Comment:
I think you could overload `processToken` like: 

```processToken(WordContext wc, Map<String, List<Location>> theMap, VariableStatementContext ctx):```

And inside this function extract text, replacement and diagnostic.

Reply:
Not exactly - it should consume different types of statement contexts, e.g. `ParagparphStatementContext` or `ErrorStatementContext`. They have a quite similar structure, but the limitations of the Java object model won't allow using it this way

Reply:
Yes, I see.